### PR TITLE
Receiving messages

### PR DESCRIPTION
**Pull request by keachi**
Read data from buffer and send it back to user space.

### CODE REVIEW

File: src/reader.c [hunk 1]
```diff
@@ -567,31 +582,44 @@ _ch_rd_read_buffer(
...
+    size_t length = 0;
     ch_message_t* msg = reader->msg;
     switch(state) {
         case CH_RD_HEADER:
             length = msg->header_len;
+            A(length == read, "msg->header_len and read are not the same");
+            msg->header = ch_alloc(read + 1);
+            for (size_t i=0; i<read; i++) {
+                msg->actor[i] = source_buf[i];
+            }
```
Comment (ganwell) [inline src/reader.c]: Please use memcpy. Or some sticklers (like me) will complain. I am actually quite indifferent about this, but I know its frowned upon.

File: src/message_etest.c [hunk 2]
```diff
@@ -93,6 +109,27 @@ ch_sent(ch_message_t* msg, int status, float load)
 
 static
 void
+ch_recv_echo_message_cb(ch_chirp_t* chirp, ch_message_t* msg)
+{
+    A(chirp->_init == CH_CHIRP_MAGIC, "Not a ch_chirp_t*");
+    A(msg != NULL, "Not a ch_message_t*");
+    if(!(msg->message_type & CH_MSG_REQ_ACK)) {
```
Comment (keachi) [inline src/message_etest.c]: I moved the callback calling code, so only ``CH_MSG_REQ_ACK`` messages will call the callback function. If more types are introduced, we must look again where to call the callback functions.

File: src/reader.c [hunk 3]
```diff
@@ -326,8 +321,18 @@ _ch_rd_handle_msg(
...
+    } else {
+        E(
+            chirp,
+            "No receiving callback function registered. "
+            "ch_chirp_t:%p",
+            (void*) chirp
+        );
+    }
+
+    /* TODO release to done
```
Comment (keachi) [inline src/reader.c]: Added that todo

File: src/reader.c [hunk 4]
```diff
@@ -356,16 +361,26 @@ _ch_rd_handle_msg(
         L(
             chirp,
             "Read message with id: %s, serial:%s from %s:%d type:%d "
-            "ch_chirp_t:%p, ch_connection_t:%p",
+            "ch_chirp_t:%p, ch_connection_t:%p, data_len:%d, data:%s",
```
Comment (keachi) [inline src/reader.c]: I changed that log message.

File: src/reader.c [hunk 5]
```diff
@@ -356,16 +361,26 @@ _ch_rd_handle_msg(
...
-            (void*) conn
+            (void*) conn,
+            msg->data_len,
+            (char*) msg->data
         );
 #   endif
+    if (msg->data_len > 0) {
+        L(
+            chirp,
+            "data_len:%d, data:%s",
```
Comment (keachi) [inline src/reader.c]: ``data_len`` before ``data`` should be right. According to #117 the length come before data.

File: src/reader.c [hunk 6]
```diff
@@ -567,31 +582,44 @@ _ch_rd_read_buffer(
...
     (void)(source_buf);
-    (void)(read);
-    int length = 0;
+    size_t length = 0;
     ch_message_t* msg = reader->msg;
     switch(state) {
         case CH_RD_HEADER:
             length = msg->header_len;
+            A(length == read, "msg->header_len and read are not the same");
+            msg->header = ch_alloc(read + 1);
```
Comment (keachi) [inline src/reader.c]: Todo is added

File: src/reader.c (hunk 1 above)
Comment (keachi) [inline src/reader.c]: I use now memcpy.

File: src/reader.c (hunk 5 above)
Comment (keachi) [inline src/reader.c]: ``data`` is now removed.

**Pull request merged**